File: erlang/grains/src/grains.erl
-module(grains).

-export([square/1, total/0, test_version/0]).

square(N) ->
  lists:foldl(fun (_, R) -> R * 2 end, 1, lists:duplicate(N - 1, nil)).

total() ->
  lists:foldl(fun (_, R) -> R * 2 end, 1, lists:duplicate(64, nil)) - 1.

test_version() -> 1.

File: haskell/grains/src/Grains.hs
module Grains (square, total) where

square :: Int -> Maybe Integer
square n
  | n <= 0 = Nothing
  | n > 64 = Nothing
  | otherwise = Just $ last $ take n $ board

total :: Integer
total = sum board

board :: [Integer]
board = take 64 $ iterate (* 2) 1
